lib/E2EEnabledPathCache.php

```diff
@@ -45,53 +47,83 @@ public function __construct() {
 	}
 
 	/**
-	 * Checks if the path is an E2E folder or inside an E2E folder
+	 * Checks if the node is an E2EE folder or inside an E2EE folder.
+     *
+     * @return bool true, if the node is valid and E2EE, false otherwise
 	 */
 	public function isE2EEnabledPath(Node $node): bool {
-		if ($node->isEncrypted()) {
-			return true;
-		}
-		$storage = $node->getStorage();
-		$cache = $storage->getCache();
-		return $this->getEncryptedStates($cache, $node, $storage);
+        // only checking for $node->isEncrypted() will lead to false-positives for SSE files,
+        // as they have the encryption flag set but are not E2E, so we need to check the folder's flag,
+        // as folders having the encryption flag set are always E2E
+        if ($node instanceof Folder && $node->isEncrypted()) {
+            return true;
+        }
+
+        try {
+            $storage = $node->getStorage();
+        } catch (NotFoundException $e) {
+            return false;
+        }
+        // if not home storage, fallback to EncryptionManager
+        if (!$storage->instanceOfStorage(IHomeStorage::class)) {
+            return EncryptionManager::isEncryptedFile($node);
+        }
+
+        // walk path backwards while caching each node's state
+		return $this->getEncryptedStates((string) $storage->getCache()->getNumericStorageId(), $node);
 	}
 
 	/**
-	 * Get the encryption state for the path
+     * Determines a node's E2E encryption state by walking up the tree. Caches each node's state on the way.
+     *
+     * @param string $storageId the node's storage id, used for caching
+     * @param Node $node the node to check
+     *
+     * @return bool true, if the node is valid and E2EE, false otherwise
 	 */
-	protected function getEncryptedStates(ICache $cache, $node, IStorage $storage): bool {
-		if (!$storage->instanceOfStorage(IHomeStorage::class)) {
-			return false;
-		}
+	protected function getEncryptedStates(string $storageId, Node $node): bool {
+
+        try {
+            $nodeId = $node->getId();
+        } catch (InvalidPathException|NotFoundException $e) {
+            return false;
+        }
 
-		$storageId = (string)$cache->getNumericStorageId();
-		if (!isset($this->perStorageEncryptedStateCache[$storageId])) {
+        // initialize array for storage id if necessary
+        if (!isset($this->perStorageEncryptedStateCache[$storageId])) {
```

Comment:
Tabs instead of spaces

```suggestion
		// initialize array for storage id if necessary
		if (!isset($this->perStorageEncryptedStateCache[$storageId])) {
```